File: python/matp/groups/manager.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Matryoshka Groups - Production Group Chat System

Combines Fractal Group Ratchet with Matryoshka steganography
for invisible multi-party encrypted communication.

License: Apache 2.0
Author: Sangeet Sharma
"""

import secrets
import json
import time
from typing import Dict, List, Optional
from .fractal_ratchet import FractalGroupRatchet
from ..protocol import MatryoshkaProtocol


class MatryoshkaGroup:
    """
    Production-ready group chat with invisible encryption.
    
    Features:
    - Multi-party encryption (Fractal Group Ratchet)
    - Invisible traffic (Matryoshka steganography)
    - Forward secrecy (per-message keys)
    - Member management (invite, remove)
    - Seed rotation (backward secrecy)
    """
    
    def __init__(self, group_id: str, group_name: str, creator_id: str):
        """
        Initialize group.
        
        Args:
            group_id: Unique group identifier
            group_name: Human-readable group name
            creator_id: User ID of group creator
        """
        self.group_id = group_id
        self.group_name = group_name
        self.creator_id = creator_id
        self.created_at = time.time()
        
        # Fractal Group Ratchet for encryption
        self.ratchet = FractalGroupRatchet()
        self.group_seed = self.ratchet.group_seed
        
        # Member management
        self.members: List[str] = [creator_id]
        self.admins: List[str] = [creator_id]
        
        # Message history (optional, for persistence)
        self.message_history: List[dict] = []
    
    def get_group_info(self) -> dict:
        """Get group metadata."""
        return {
            "group_id": self.group_id,
            "group_name": self.group_name,
            "creator": self.creator_id,
            "members": self.members,
            "admins": self.admins,
            "member_count": len(self.members),
            "fingerprint": self.ratchet.get_fingerprint(),
            "created_at": self.created_at
        }
    
    def add_member(self, user_id: str, is_admin: bool = False):
        """Add member to group."""
        if user_id not in self.members:
            self.members.append(user_id)
            if is_admin:
                self.admins.append(user_id)
    
    def remove_member(self, user_id: str, requester_id: str) -> bool:
        """
        Remove member from group (admin only).
        
        Returns:
            bool: True if removed, False if not authorized
        """
        if requester_id not in self.admins:
            return False
        
        if user_id in self.members:
            self.members.remove(user_id)
            if user_id in self.admins:
                self.admins.remove(user_id)
        
        return True
    
    def send_message(self, sender_id: str, message: str, 
                    use_steganography: bool = True) -> dict:
        """
        Send message to group (invisible!).
        
        Args:
            sender_id: User ID of sender
            message: Message text
            use_steganography: Hide in cover traffic (default: True)
        
        Returns:
            dict: Invisible message envelope
        """
        # Encrypt with Fractal Group Ratchet
        encrypted_envelope = self.ratchet.encrypt_for_group(message)
        
        # Create group message
        group_message = {
            "type": "group_message",
            "group_id": self.group_id,
            "sender": sender_id,
            "timestamp": time.time(),
            "encrypted": encrypted_envelope
        }
        
        # Store in history
        self.message_history.append({
            "sender": sender_id,
            "timestamp": group_message["timestamp"],
            "layer": encrypted_envelope["layer"]
        })
        
        # Hide in steganography (invisible!)
        if use_steganography:
            # Embed group message in cover traffic
            import base64
            payload = base64.b64encode(json.dumps(group_message).encode()).decode()
            
            cover = {
                "status": "success",
                "data": {
                    "user_id": 12345,
                    "session_token": payload,  # Hidden group message
                    "preferences": {"theme": "dark", "lang": "en"},
                    "timestamp": int(time.time())
                },
                "meta": {"version": "2.1.0", "server": "api-01"}
            }
            return cover
        else:
            return group_message
    
    def receive_message(self, ghost_msg_data: dict) -> dict:
        """
        Receive and decrypt group message.
        
        Args:
            ghost_msg_data: Invisible message data
        
        Returns:
            dict: Decrypted message with metadata
        """
        # Extract from steganography cover
        # The message is in the session_token field
        if "data" in ghost_msg_data and "session_token" in ghost_msg_data["data"]:
            encoded = ghost_msg_data["data"]["session_token"]
            import base64
            encrypted = base64.b64decode(encoded)
            
            # Decrypt the outer envelope (just base64 decode for now)
            decrypted_json = encrypted.decode('utf-8')
        else:
            decrypted_json = json.dumps(ghost_msg_data)
        
        group_message = json.loads(decrypted_json)
        
        # Verify group ID
        if group_message["group_id"] != self.group_id:
            raise ValueError("Message not for this group")
        
        # Decrypt with Fractal Group Ratchet
        plaintext = self.ratchet.decrypt_from_group(
            group_message["encrypted"]
        )
        
        return {
            "sender": group_message["sender"],
            "message": plaintext,
            "timestamp": group_message["timestamp"],
            "group_id": self.group_id,
            "group_name": self.group_name
        }
    
    def export_invite(self, from_layer: Optional[int] = None) -> dict:
        """
        Export group invite for new member.
        
        Args:
            from_layer: Message index to start from (None = current)
        
        Returns:
            dict: Invite data to share with new member
        """
        if from_layer is None:
            from_layer = self.ratchet.message_counter
        
        session = self.ratchet.export_session(from_layer=from_layer)
        
        return {
            "type": "group_invite",
            "group_id": self.group_id,
            "group_name": self.group_name,
            "creator": self.creator_id,
            "members": self.members,
            "session": session,
            "invited_at": time.time()
        }
    
    def rotate_group_seed(self, requester_id: str) -> Optional[bytes]:
        """
        Rotate group seed (admin only, for backward secrecy).
        
        Args:
            requester_id: User requesting rotation
        
        Returns:
            bytes: New seed if authorized, None otherwise
        """
        if requester_id not in self.admins:
            return None
        
        new_seed = self.ratchet.rotate_seed()
        self.group_seed = new_seed
        return new_seed


class MatryoshkaGroupManager:
    """
    Manage multiple groups for a user.
    
    Production-ready group chat manager with:
    - Multiple group support
    - 1-to-1 encrypted invites
    - Invisible group messages
    - Member management
    """
    
    def __init__(self, user_id: str):
        """
        Initialize group manager.
        
        Args:
            user_id: Unique user identifier
        """
        self.user_id = user_id
        self.groups: Dict[str, MatryoshkaGroup] = {}
        
        # For 1-to-1 encrypted invites
        self.private_key, self.public_key = MatryoshkaProtocol.generate_keypair()
        self.peer_sessions: Dict[str, MatryoshkaProtocol] = {}
    
    def create_group(self, group_id: str, group_name: str) -> MatryoshkaGroup:
        """
        Create new group.
        
        Args:
            group_id: Unique group identifier
            group_name: Human-readable name
        
        Returns:
            MatryoshkaGroup: Created group
        """
        group = MatryoshkaGroup(group_id, group_name, self.user_id)
        self.groups[group_id] = group
        return group
    
    def join_group(self, invite_data: dict) -> MatryoshkaGroup:
        """
        Join group from invite.
        
        Args:
            invite_data: Invite exported by group admin
        
        Returns:
            MatryoshkaGroup: Joined group
        """
        # Create group
        group = MatryoshkaGroup(
            invite_data["group_id"],
            invite_data["group_name"],
            invite_data["creator"]
        )
        
        # Import session
        group.ratchet.import_session(invite_data["session"])
        group.group_seed = group.ratchet.group_seed
        
        # Set members
        group.members = invite_data["members"]
        group.members.append(self.user_id)
        
        # Store group
        self.groups[group.group_id] = group
        
        return group
    
    def send_to_group(self, group_id: str, message: str) -> dict:
        """
        Send invisible message to group.
        
        Args:
            group_id: Target group ID
            message: Message text
        
        Returns:
            dict: Invisible message (looks like normal web traffic)
        """
        if group_id not in self.groups:
            raise ValueError(f"Not in group: {group_id}")
        
        group = self.groups[group_id]
        return group.send_message(self.user_id, message)
    
    def receive_group_message(self, ghost_msg_data: dict) -> dict:
        """
        Receive and decrypt group message.
        
        Args:
            ghost_msg_data: Invisible message data
        
        Returns:
            dict: Decrypted message with metadata
        """
        # Try each group until we find the right one
        for group in self.groups.values():
            try:
                return group.receive_message(ghost_msg_data)
            except (ValueError, KeyError):
                continue
        
        raise ValueError("Message not for any of your groups")
    
    def invite_to_group(self, group_id: str, invitee_id: str, 
                       invitee_public_key) -> dict:
        """
        Invite user to group (encrypted 1-to-1).
        
        Args:
            group_id: Group to invite to
            invitee_id: User to invite
            invitee_public_key: Invitee's X25519 public key
        
        Returns:
            dict: Encrypted invite message
        """
        if group_id not in self.groups:
            raise ValueError(f"Not in group: {group_id}")
        
        group = self.groups[group_id]
        
        # Create 1-to-1 session
        shared_secret = MatryoshkaProtocol.derive_shared_secret(
            self.private_key,
            invitee_public_key
        )
        peer_session = MatryoshkaProtocol(key=shared_secret)
        self.peer_sessions[invitee_id] = peer_session
        
        # Export invite
        invite = group.export_invite()
        
        # Send encrypted (invisible!)
        invite_msg = peer_session.send_message(
            json.dumps(invite),
            use_steganography=True
        )
        
        # Add to group
        group.add_member(invitee_id)
        
        return invite_msg.cover_data
    
    def get_all_groups(self) -> List[dict]:
        """Get info for all groups."""
        return [group.get_group_info() for group in self.groups.values()]
    
    def leave_group(self, group_id: str):
        """Leave group."""
        if group_id in self.groups:
            del self.groups[group_id]


# Production usage example
if __name__ == "__main__":
    print("=== Matryoshka Groups - Production Demo ===\n")
    
    # Create users
    print("1. Creating users...")
    alice = MatryoshkaGroupManager(user_id="alice")
    bob = MatryoshkaGroupManager(user_id="bob")
    charlie = MatryoshkaGroupManager(user_id="charlie")
    print("   ✓ Alice, Bob, Charlie created\n")
    
    # Alice creates group
    print("2. Alice creates 'Secret Project' group...")
    group = alice.create_group("project-x", "Secret Project")
    print(f"   ✓ Group created: {group.group_id}")
    print(f"   ✓ Fingerprint: {group.ratchet.get_fingerprint()}\n")
    
    # Alice invites Bob and Charlie
    print("3. Alice invites Bob and Charlie...")
    bob_invite = alice.invite_to_group("project-x", "bob", bob.public_key)
    charlie_invite = alice.invite_to_group("project-x", "charlie", charlie.public_key)
    print("   ✓ Invites sent (encrypted 1-to-1)\n")
    
    # Bob and Charlie join (in real app, they'd receive and decrypt invites)
    print("4. Bob and Charlie join group...")
    invite_data = group.export_invite()
    bob.join_group(invite_data)
    charlie.join_group(invite_data)
    print("   ✓ Bob joined")
    print("   ✓ Charlie joined\n")
    
    # Alice sends invisible message
    print("5. Alice sends invisible message...")
    msg = alice.send_to_group("project-x", "Meeting at 3pm tomorrow!")
    print("   ✓ Message sent (looks like normal API traffic)")
    print(f"   ✓ Cover: {list(msg.keys())}\n")
    
    # Bob and Charlie receive
    print("6. Bob and Charlie receive message...")
    bob_msg = bob.receive_group_message(msg)
    charlie_msg = charlie.receive_group_message(msg)
    print(f"   Bob sees: '{bob_msg['message']}'")
    print(f"   Charlie sees: '{charlie_msg['message']}'\n")
    
    # Group info
    print("7. Group information...")
    info = alice.groups["project-x"].get_group_info()
    print(f"   Name: {info['group_name']}")
    print(f"   Members: {info['members']}")
    print(f"   Member count: {info['member_count']}\n")
    
    print("=== Demo Complete ===")
    print("\n Production-ready group chat with invisible encryption!")

File: rust/core/src/groups/manager.rs
//! # Matryoshka Groups - Production Group Chat System
//!
//! Combines Fractal Group Ratchet with Matryoshka steganography
//! for invisible multi-party encrypted communication.

use std::collections::HashMap;
use serde::{Serialize, Deserialize};
use base64::{engine::general_purpose, Engine as _};

use super::{FractalGroupRatchet, GroupError};
use super::fractal_ratchet::{GroupEnvelope, SessionExport};

/// Production-ready group chat with invisible encryption
pub struct MatryoshkaGroup {
    group_id: String,
    group_name: String,
    creator_id: String,
    created_at: f64,
    
    ratchet: FractalGroupRatchet,
    
    members: Vec<String>,
    admins: Vec<String>,
    
    message_history: Vec<MessageRecord>,
}

#[derive(Debug, Clone)]
#[allow(dead_code)]
struct MessageRecord {
    sender: String,
    timestamp: f64,
    layer: u32,
}

impl MatryoshkaGroup {
    /// Create new group
    pub fn new(group_id: String, group_name: String, creator_id: String) -> Self {
        Self {
            group_id,
            group_name,
            creator_id: creator_id.clone(),
            created_at: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_secs_f64(),
            ratchet: FractalGroupRatchet::new(None),
            members: vec![creator_id.clone()],
            admins: vec![creator_id],
            message_history: Vec::new(),
        }
    }
    
    /// Get group metadata
    pub fn get_group_info(&self) -> GroupInfo {
        GroupInfo {
            group_id: self.group_id.clone(),
            group_name: self.group_name.clone(),
            creator: self.creator_id.clone(),
            members: self.members.clone(),
            admins: self.admins.clone(),
            member_count: self.members.len(),
            fingerprint: self.ratchet.get_fingerprint().to_string(),
            created_at: self.created_at,
        }
    }
    
    /// Add member to group
    pub fn add_member(&mut self, user_id: String, is_admin: bool) {
        if !self.members.contains(&user_id) {
            self.members.push(user_id.clone());
            if is_admin {
                self.admins.push(user_id);
            }
        }
    }
    
    /// Remove member from group (admin only)
    pub fn remove_member(&mut self, user_id: &str, requester_id: &str) -> Result<(), GroupError> {
        if !self.admins.contains(&requester_id.to_string()) {
            return Err(GroupError::PermissionDenied("Not an admin".to_string()));
        }
        
        self.members.retain(|m| m != user_id);
        self.admins.retain(|a| a != user_id);
        
        Ok(())
    }
    
    /// Send message to group (invisible!)
    pub fn send_message(&mut self, sender_id: &str, message: &str, use_steganography: bool) 
        -> Result<Vec<u8>, GroupError> 
    {
        let encrypted_envelope = self.ratchet.encrypt_for_group(message)?;
        
        let group_message = GroupMessage {
            msg_type: "group_message".to_string(),
            group_id: self.group_id.clone(),
            sender: sender_id.to_string(),
            timestamp: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_secs_f64(),
            encrypted: encrypted_envelope.clone(),
        };
        
        // Store in history
        self.message_history.push(MessageRecord {
            sender: sender_id.to_string(),
            timestamp: group_message.timestamp,
            layer: encrypted_envelope.layer,
        });
        
        if use_steganography {
            // Embed in cover traffic
            let payload = serde_json::to_string(&group_message)
                .map_err(|e| GroupError::EncryptionError(e.to_string()))?;
            let encoded = general_purpose::STANDARD.encode(payload);
            
            let cover = serde_json::json!({
                "status": "success",
                "data": {
                    "user_id": 12345,
                    "session_token": encoded,
                    "preferences": {"theme": "dark", "lang": "en"},
                    "timestamp": group_message.timestamp as u64
                },
                "meta": {"version": "2.1.0", "server": "api-01"}
            });
            
            serde_json::to_vec(&cover)
                .map_err(|e| GroupError::EncryptionError(e.to_string()))
        } else {
            serde_json::to_vec(&group_message)
                .map_err(|e| GroupError::EncryptionError(e.to_string()))
        }
    }
    
    /// Receive and decrypt group message
    pub fn receive_message(&mut self, ghost_data: &[u8]) -> Result<ReceivedMessage, GroupError> {
        // Try to parse as JSON
        let json_data: serde_json::Value = serde_json::from_slice(ghost_data)
            .map_err(|e| GroupError::DecryptionError(e.to_string()))?;
        
        // Extract from steganography if present
        let group_message: GroupMessage = if let Some(data) = json_data.get("data") {
            if let Some(token) = data.get("session_token") {
                let encoded = token.as_str()
                    .ok_or_else(|| GroupError::DecryptionError("Invalid token".to_string()))?;
                let decoded = general_purpose::STANDARD.decode(encoded)
                    .map_err(|e| GroupError::DecryptionError(e.to_string()))?;
                serde_json::from_slice(&decoded)
                    .map_err(|e| GroupError::DecryptionError(e.to_string()))?
            } else {
                serde_json::from_value(json_data)
                    .map_err(|e| GroupError::DecryptionError(e.to_string()))?
            }
        } else {
            serde_json::from_value(json_data)
                .map_err(|e| GroupError::DecryptionError(e.to_string()))?
        };
        
        // Verify group ID
        if group_message.group_id != self.group_id {
            return Err(GroupError::DecryptionError("Message not for this group".to_string()));
        }
        
        // Decrypt with ratchet
        let plaintext = self.ratchet.decrypt_from_group(&group_message.encrypted)?;
        
        Ok(ReceivedMessage {
            sender: group_message.sender,
            message: plaintext,
            timestamp: group_message.timestamp,
            group_id: self.group_id.clone(),
            group_name: self.group_name.clone(),
        })
    }
    
    /// Export invite for new member
    pub fn export_invite(&self, from_layer: Option<u32>) -> GroupInvite {
        let layer = from_layer.unwrap_or(self.ratchet.message_counter);
        let session = self.ratchet.export_session(layer);
        
        GroupInvite {
            invite_type: "group_invite".to_string(),
            group_id: self.group_id.clone(),
            group_name: self.group_name.clone(),
            creator: self.creator_id.clone(),
            members: self.members.clone(),
            session,
            invited_at: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_secs_f64(),
        }
    }
    
    /// Rotate group seed (admin only)
    pub fn rotate_group_seed(&mut self, requester_id: &str) -> Result<[u8; 32], GroupError> {
        if !self.admins.contains(&requester_id.to_string()) {
            return Err(GroupError::PermissionDenied("Not an admin".to_string()));
        }
        
        Ok(self.ratchet.rotate_seed())
    }
}

/// Group metadata
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GroupInfo {
    pub group_id: String,
    pub group_name: String,
    pub creator: String,
    pub members: Vec<String>,
    pub admins: Vec<String>,
    pub member_count: usize,
    pub fingerprint: String,
    pub created_at: f64,
}

/// Group message structure
#[derive(Debug, Clone, Serialize, Deserialize)]
struct GroupMessage {
    #[serde(rename = "type")]
    msg_type: String,
    group_id: String,
    sender: String,
    timestamp: f64,
    encrypted: GroupEnvelope,
}

/// Received message
#[derive(Debug, Clone)]
pub struct ReceivedMessage {
    pub sender: String,
    pub message: String,
    pub timestamp: f64,
    pub group_id: String,
    pub group_name: String,
}

/// Group invite
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GroupInvite {
    #[serde(rename = "type")]
    invite_type: String,
    pub group_id: String,
    pub group_name: String,
    pub creator: String,
    pub members: Vec<String>,
    pub session: SessionExport,
    pub invited_at: f64,
}

/// Manage multiple groups for a user
pub struct MatryoshkaGroupManager {
    user_id: String,
    groups: HashMap<String, MatryoshkaGroup>,
}

impl MatryoshkaGroupManager {
    /// Create new group manager
    pub fn new(user_id: String) -> Self {
        Self {
            user_id,
            groups: HashMap::new(),
        }
    }
    
    /// Create new group
    pub fn create_group(&mut self, group_id: String, group_name: String) -> &mut MatryoshkaGroup {
        let group = MatryoshkaGroup::new(group_id.clone(), group_name, self.user_id.clone());
        self.groups.insert(group_id.clone(), group);
        self.groups.get_mut(&group_id).unwrap()
    }
    
    /// Join group from invite
    pub fn join_group(&mut self, invite: &GroupInvite) -> Result<(), GroupError> {
        let mut group = MatryoshkaGroup::new(
            invite.group_id.clone(),
            invite.group_name.clone(),
            invite.creator.clone(),
        );
        
        group.ratchet.import_session(&invite.session)?;
        group.members = invite.members.clone();
        group.members.push(self.user_id.clone());
        
        self.groups.insert(invite.group_id.clone(), group);
        Ok(())
    }
    
    /// Send message to group
    pub fn send_to_group(&mut self, group_id: &str, message: &str) -> Result<Vec<u8>, GroupError> {
        let group = self.groups.get_mut(group_id)
            .ok_or_else(|| GroupError::GroupNotFound(group_id.to_string()))?;
        
        group.send_message(&self.user_id, message, true)
    }
    
    /// Receive group message
    pub fn receive_group_message(&mut self, ghost_data: &[u8]) -> Result<ReceivedMessage, GroupError> {
        for group in self.groups.values_mut() {
            if let Ok(msg) = group.receive_message(ghost_data) {
                return Ok(msg);
            }
        }
        
        Err(GroupError::DecryptionError("Message not for any group".to_string()))
    }
    
    /// Get all groups
    pub fn get_all_groups(&self) -> Vec<GroupInfo> {
        self.groups.values().map(|g| g.get_group_info()).collect()
    }
    
    /// Leave group
    pub fn leave_group(&mut self, group_id: &str) {
        self.groups.remove(group_id);
    }
    
    /// Get specific group
    pub fn get_group(&self, group_id: &str) -> Option<&MatryoshkaGroup> {
        self.groups.get(group_id)
    }
    
    /// Get mutable group reference
    pub fn get_group_mut(&mut self, group_id: &str) -> Option<&mut MatryoshkaGroup> {
        self.groups.get_mut(group_id)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_group_creation() {
        let mut alice = MatryoshkaGroupManager::new("alice".to_string());
        let group = alice.create_group("project-x".to_string(), "Secret Project".to_string());
        
        let info = group.get_group_info();
        assert_eq!(info.group_id, "project-x");
        assert_eq!(info.members.len(), 1);
    }
    
    #[test]
    fn test_group_messaging() {
        let mut alice = MatryoshkaGroupManager::new("alice".to_string());
        let mut bob = MatryoshkaGroupManager::new("bob".to_string());
        
        // Alice creates group
        alice.create_group("test-group".to_string(), "Test".to_string());
        
        // Export invite
        let invite = alice.get_group("test-group").unwrap().export_invite(None);
        
        // Bob joins
        bob.join_group(&invite).unwrap();
        
        // Alice sends message
        let msg = alice.send_to_group("test-group", "Hello Bob!").unwrap();
        
        // Bob receives
        let received = bob.receive_group_message(&msg).unwrap();
        assert_eq!(received.message, "Hello Bob!");
        assert_eq!(received.sender, "alice");
    }
}
